src/test/scala/chiselTests/ModuleChoiceSpec.scala

```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chiselTests
+
+import chisel3._
+import chiselTests.{ChiselFlatSpec, MatchesAndOmits, Utils}
+import _root_.circt.stage.ChiselStage
+
+object Platform extends option.Option {
+  object FPGA extends option.Case
+  object ASIC extends option.Case
+}
+
+object Performance extends option.Option {
+  object Fast extends option.Case
+  object Small extends option.Case
+}
+
+class TargetIO extends Bundle {
+  val in = Flipped(UInt(8.W))
+  val out = UInt(8.W)
+}
+
+class FPGATarget extends FixedIOExtModule[TargetIO](new TargetIO) {}
+
+class ASICTarget extends FixedIOExtModule[TargetIO](new TargetIO) {}
+
+class VerifTarget extends FixedIORawModule[TargetIO](new TargetIO) {}
+
+class ModuleChoiceSpec extends ChiselFlatSpec with Utils with MatchesAndOmits {
+  it should "emit options and cases" in {
+    class ModuleWithChoice extends Module {
+      val out = IO(UInt(8.W))
+
+      val inst =
+        ModuleChoice(new VerifTarget)(Seq(Platform.FPGA -> new FPGATarget, Platform.ASIC -> new ASICTarget))
```

Comment:
Is this something that should have to be specified at every call site, or is this more of a property of the type of the default choice? I'm envisioning some sort of typeclass API, eg.

```scala
// ...
object ModuleChoice {
  def apply[T <: Data : ModuleChoices](default: => FixedIOBaseModule[T]): T = ???
}
// ...

class FPGATarget extends FixedIOExtModule[TargetIO](new TargetIO) {}

class ASICTarget extends FixedIOExtModule[TargetIO](new TargetIO) {}

class VerifTarget extends FixedIORawModule[TargetIO](new TargetIO) {}
object VerifTarget {
  implicit val choices = Choices[TargetIO](Platform.FPGA -> new FPGATarget, Platform.ASIC -> new ASICTarget))
}
// ...
val inst = ModuleChoice(new VerifTarget)
```
Then of course if there are different choices for different call sites the user can provide the argument explicitly using normal typeclass pattern mechanisms.

Reply:
This feature should be used at call sites. Users should feel free to add wrappers to group some choices, but that's not something that the Chisel feature should bake in.

Reply:
But why? You're already opting in at the call site by using `ModuleChoice` instead of `Module`. The question is how often for a given type do you expect the user to use different choices or is it probably basically always going to be the same? Yes the user can write a function, but making the default behavior more elegant in the common case is still generally a good idea.

Reply:
I was a bit confused by the code you sketched up since it hides the variants from the default.

Reply:
> I was a bit confused by the code you sketched up since it hides the variants from the default

What I wrote is just how you express typeclasses in Scala 2, by virtue of putting the variants in the companion object for the default, it's very much associated with the default.

I agree it's fine to go with the more basic API for the time being, although my intuition is still that the API people end up wanting is something closer to the typeclass-y API and it will be slightly unfortunate that we have to give the higher-level abstraction a different name. We can cross that bridge later though.